apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/SamhsaUtil.java

```diff
@@ -153,6 +225,210 @@ private List<McsTag> checkAndProcessMcsClaim(RdaMcsClaim mcsClaim) {
     return Collections.emptyList();
   }
 
+  /**
+   * Checks for SAMHSA codes in an HHA claim and constructs the tags.
+   *
+   * @param claim The claim to check.
+   * @return A list of tag entities to persist.
+   */
+  private List<HhaTag> checkAndProcessHhaClaim(HHAClaim claim)
```

Comment:
I think this logic can operate on an input of `SamhsaAdapterBase` so it doesn't need to be duplicated for each claim type.

Reply:
Maybe? Each tag type would still need to be instantiated individually, so I'm not sure how much value there would be to moving this to the adapter. Can you elaborate on what you're thinking?

Reply:
Yeah, we would probably need a method in the adapter called `buildTag` or something in the adapter that instantiates the tag object from the claim ID, code, and details, but if we did that, we wouldn't need a separate `checkAndProcessXXXClaim` method for each claim type in here and we could get rid of the big switch statement on line 135. I haven't tried it myself, so I could be missing something, but I think that would work?

Reply:
I'm hoping the adapter can simplify the logic in this file as much as possible so we can eliminate all of the switch statements and confine any duplicated code to the adapters. It won't prevent all the annoying boilerplate, but it at least prevents the need for big switch statements and keeps the main logic separate which I think helps with readability.

Reply:
We would still need to know which adapter to instantiate; we may be able to move the switch logic to a factory class, but I don't think it will be possible to eliminate it completely.

Reply:
Yeah that's true. It's just one line to instantiate the adapter though so that's not the end of the world.

Reply:
Refactored based on this conversation.